File: ios/RNADBMobileAnalytics.m
#import "RNADBMobileAnalytics.h"
#import "ADBMobile.h"
#import <CoreLocation/CoreLocation.h>

@implementation RNADBMobileAnalytics

RCT_EXPORT_MODULE();

+ (BOOL)requiresMainQueueSetup
{
    return YES; // reason: constantsToExport
}

- (NSDictionary *)constantsToExport
{
    return @{@"PROXIMITY_IMMEDIATE": @"PROXIMITY_IMMEDIATE",
             @"PROXIMITY_NEAR": @"PROXIMITY_NEAR",
             @"PROXIMITY_FAR": @"PROXIMITY_FAR",
             @"PROXIMITY_UNKNOWN": @"PROXIMITY_UNKNOWN"};
}

RCT_EXPORT_METHOD(trackState:(nonnull NSString *)state contextData:(NSDictionary *)contextData)
{
    [ADBMobile trackState:state data:contextData];
}

RCT_EXPORT_METHOD(trackAction:(nonnull NSString *)action contextData:(NSDictionary *)contextData) {
    [ADBMobile trackAction:action data:contextData];
}

RCT_EXPORT_METHOD(trackLocation:(NSDictionary *)coordinatesSpec contextData:(NSDictionary *)contextData)
{
    id objLat = coordinatesSpec[@"latitude"];
    id objLon = coordinatesSpec[@"longitude"];
    if (!objLat || !objLon) {
        NSLog(@"must provide latitude and longitude!");
        return;
    }
    double lat = [objLat doubleValue];
    double lon = [objLon doubleValue];

    id objAlt = coordinatesSpec[@"altitude"];
    double alt = objAlt ? [objAlt doubleValue] : 0;

    double vAccuracy = objAlt ? [coordinatesSpec[@"altitudeAccuracy"] doubleValue] : -1;
    double hAccuracy = [coordinatesSpec[@"accuracy"] doubleValue];
    double course = [coordinatesSpec[@"heading"] doubleValue];
    double speed = [coordinatesSpec[@"speed"] doubleValue];

    CLLocationCoordinate2D coord2d;
    coord2d.latitude = lat;
    coord2d.longitude = lon;

    CLLocation *location = [[CLLocation alloc]
                            initWithCoordinate:coord2d
                            altitude:alt
                            horizontalAccuracy:hAccuracy
                            verticalAccuracy:vAccuracy
                            course:course
                            speed:speed
                            timestamp:[NSDate date]];
    [ADBMobile trackLocation:location data:contextData];
}

RCT_EXPORT_METHOD(trackBeacon:(nonnull NSString *)uuid major:(nonnull NSString *)major minor:(nonnull NSString *)minor proximityName:(nonnull NSString *)proximityName contextData:(NSDictionary *)contextData)
{
    NSLog(@"trackBeacon() not supported on iOS. Meanwhile you can call\n\n"
          "  [ADBMobile trackBeacon:beacon data: contextData];\n\n"
          "From one of the CLLocationManager delegate methods, see:\n"
          "https://developer.apple.com/documentation/corelocation/determining_the_proximity_to_an_ibeacon");
}

RCT_EXPORT_METHOD(clearBeacon)
{
    [ADBMobile trackingClearCurrentBeacon];
}

RCT_EXPORT_METHOD(trackLifetimeValueIncrease:(double)amount contextData:(NSDictionary *)contextData)
{
    NSDecimalNumber *d = [[NSDecimalNumber alloc] initWithDouble:amount];
    [ADBMobile trackLifetimeValueIncrease:d data:contextData];
}

RCT_EXPORT_METHOD(trackTimedActionStart:(nonnull NSString *)action contextData:(NSDictionary *)contextData)
{
    [ADBMobile trackTimedActionStart:action data:contextData];
}

RCT_EXPORT_METHOD(trackTimedActionUpdate:(nonnull NSString *)action contextData:(NSDictionary *)contextData)
{
    [ADBMobile trackTimedActionUpdate:action data:contextData];
}

RCT_EXPORT_METHOD(trackTimedActionEnd:(nonnull NSString *)action hit:(BOOL)hit)
{
    [ADBMobile trackTimedActionEnd:action logic:^BOOL (NSTimeInterval inAppDuration, NSTimeInterval totalDuration, NSMutableDictionary *data) {
        return hit;
    }];
}

RCT_EXPORT_METHOD(trackingTimedActionExists:(nonnull NSString *)action resolver:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter)
{
    resolver([NSNumber numberWithBool:[ADBMobile trackingTimedActionExists:action]]);
}

RCT_EXPORT_METHOD(getTrackingIdentifier:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter)
{
    resolver([ADBMobile trackingIdentifier]);
}

RCT_EXPORT_METHOD(getQueueSize:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    resolver([NSNumber numberWithUnsignedInteger:[ADBMobile trackingGetQueueSize]]);
}

RCT_EXPORT_METHOD(clearQueue)
{
    [ADBMobile trackingClearQueue];
}

RCT_EXPORT_METHOD(sendQueuedHits)
{
    [ADBMobile trackingSendQueuedHits];
}

@end

File: android/src/main/java/co/work/rnadbmobile/RNADBMobileAnalytics.kt
package co.work.rnadbmobile

import com.adobe.mobile.Config
import com.adobe.mobile.Analytics

import android.location.Location

import com.facebook.common.logging.FLog
import com.facebook.react.bridge.*

class RNADBMobileAnalytics(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext), LifecycleEventListener {
    private val _tag = "RNADBMobileAnalytics"

    override fun getName() = _tag

    override fun getConstants(): Map<String, Any> = hashMapOf(
            "PROXIMITY_IMMEDIATE" to "PROXIMITY_IMMEDIATE",
            "PROXIMITY_NEAR" to "PROXIMITY_NEAR",
            "PROXIMITY_FAR" to "PROXIMITY_FAR",
            "PROXIMITY_UNKNOWN" to "PROXIMITY_UNKNOWN"
        )

    init {
        reactContext.addLifecycleEventListener(this)
        FLog.i(_tag, "added lifecycle event listener")
    }

    override fun onHostResume() = Config.collectLifecycleData(currentActivity)
    override fun onHostPause() =  Config.pauseCollectingLifecycleData()
    override fun onHostDestroy() {}

    @ReactMethod
    fun trackState(state: String, contextData: ReadableMap? = null) {
        Analytics.trackState(state, contextData?.toHashMap())
    }

    @ReactMethod
    fun trackAction(action: String, contextData: ReadableMap? = null) {
        Analytics.trackAction(action, contextData?.toHashMap())
    }

    @ReactMethod
    fun trackAdobeDeepLink(url: String) {}

    @ReactMethod
    fun trackLocation(coordinates: ReadableMap, contextData: ReadableMap? = null) {
        val location = Location("")

        location.latitude = coordinates.getDouble("latitude")
        location.longitude = coordinates.getDouble("longitude")

        if (coordinates.hasKey("altitude")) {
            location.altitude = coordinates.getDouble("altitude")
            if (android.os.Build.VERSION.SDK_INT >= 26 && coordinates.hasKey("altitudeAccuracy")) {
                location.verticalAccuracyMeters = coordinates.getDouble("altitudeAccuracy").toFloat()
            }
        }

        if (coordinates.hasKey("accuracy")) {
            location.accuracy = coordinates.getDouble("accuracy").toFloat()
        }

        if (coordinates.hasKey("heading")) {
            location.bearing = coordinates.getDouble("heading").toFloat()
        }

        if (coordinates.hasKey("speed")) {
            location.speed = coordinates.getDouble("speed").toFloat()
        }

        Analytics.trackLocation(location, contextData?.toHashMap())
    }

    @ReactMethod
    fun trackBeacon(uuid: String, major: String, minor: String, proximityName: String, contextData: ReadableMap? = null) {
        val proximity: Analytics.BEACON_PROXIMITY
        when (proximityName) {
            "PROXIMITY_IMMEDIATE" -> proximity = Analytics.BEACON_PROXIMITY.PROXIMITY_IMMEDIATE
            "PROXIMITY_NEAR" -> proximity = Analytics.BEACON_PROXIMITY.PROXIMITY_NEAR
            "PROXIMITY_FAR" -> proximity = Analytics.BEACON_PROXIMITY.PROXIMITY_FAR
            else -> proximity = Analytics.BEACON_PROXIMITY.PROXIMITY_UNKNOWN
        }
        Analytics.trackBeacon(uuid, major, minor, proximity, contextData?.toHashMap())
    }

    @ReactMethod
    fun clearBeacon() = Analytics.clearBeacon()

    @ReactMethod
    fun trackLifetimeValueIncrease(amount: Double, contextData: ReadableMap? = null) {
        Analytics.trackLifetimeValueIncrease(amount.toBigDecimal(), contextData?.toHashMap())
    }

    @ReactMethod
    fun trackTimedActionStart(action: String, contextData: ReadableMap? = null) {
        Analytics.trackTimedActionStart(action, contextData?.toHashMap())
    }

    @ReactMethod
    fun trackTimedActionUpdate(action: String, contextData: ReadableMap? = null) {
        Analytics.trackTimedActionUpdate(action, contextData?.toHashMap())
    }

    @ReactMethod
    fun trackTimedActionEnd(action: String, hit: Boolean) {
        val logic = { _: Long, _: Long, _: Map<String, Any> -> hit }
        Analytics.trackTimedActionEnd(action, logic)
    }

    @ReactMethod
    fun trackingTimedActionExists(action: String, promise: Promise) = promise.resolve(Analytics.trackingTimedActionExists(action))

    @ReactMethod
    fun getTrackingIdentifier(promise: Promise) = promise.resolve(Analytics.getTrackingIdentifier())

    @ReactMethod
    fun getQueueSize(promise: Promise) = promise.resolve(Analytics.getQueueSize().toDouble())

    @ReactMethod
    fun clearQueue() = Analytics.clearQueue()

    @ReactMethod
    fun sendQueuedHits() = Analytics.sendQueuedHits()
}
